fundraiser/src/main.sw

```diff
@@ -0,0 +1,438 @@
+    contract_id::ContractId,
+    identity::*,
+    logging::log,
+    result::*,
+    revert::revert,
+    storage::StorageMap,
+    token::transfer,
+};
+
+use abi::Fundraiser;
+use data_structures::{Campaign, CampaignInfo, Pledge};
+use errors::{CampaignError, CreationError, UserError};
+use events::{CancelledCampaignEvent, ClaimedEvent, CreatedCampaignEvent, PledgedEvent, UnpledgedEvent};
+use utils::{sender_identity, validate_id};
+
+storage {
+    /// The total number of unique campaigns that a user has created
+    /// This should only be incremented
+    /// Cancelling / Claiming should not affect this number
+    campaign_count: StorageMap<Identity,
+    u64>, /// Campaigns that have been created by a user
+    /// Map(Identity => Map(1...campaign_count => Campaign)
+    campaign_history: StorageMap<(Identity,
+    u64), Campaign>, /// Data describing the content of a campaign
+    /// Map(Campaign ID => CampaignInfo)
+    campaign_info: StorageMap<u64,
+    CampaignInfo>, /// The total number of unique campaigns that a user has pledged to
+    /// This should only be incremented.
+    /// Unpledging should not affect this number
+    pledge_count: StorageMap<Identity,
+    u64>, /// Record of how much a user has pledged to a specific campaign
+    /// Locked after the deadline
+    /// Map(Identity => Map(1...pledge_count => Pledge))
+    pledge_history: StorageMap<(Identity,
+    u64), Pledge>, /// O(1) look-up to prevent iterating over pledge_history
+    /// Map(Identity => Map(Campaign ID => Pledge History Index))
+    pledge_history_index: StorageMap<(Identity,
+    u64), u64>, /// The number of campaigns created by all users
+    total_campaigns: u64,
+}
+
+impl Fundraiser for Contract {
+    /// Creates a data structure representing a campaign that users can pledge to
+    ///
+    /// Instead of having a contract per campaign we create an internal representation for the data
+    /// and manage it via mappings.
+    ///
+    /// # Arguments
+    ///
+    /// * `asset` - A coin that the campaign accepts as a pledge
+    /// * `beneficiary` - The recipient to whom the pledge will be sent to upon a successful campaign
+    /// * `deadline` - Block height used to dictate the end time of a campaign
+    /// * `target_amount` - The amount of `asset` required to deem the campaign a success
+    ///
+    /// # Reverts
+    ///
+    /// * When `asset` is the BASE_ASSET
+    /// * When the `deadline` is not ahead of the current block height
+    /// * When the `target_amount` is 0
+    /// * When an AuthError is generated
+    #[storage(read, write)]fn create_campaign(asset: ContractId, beneficiary: Identity, deadline: u64, target_amount: u64) {
+        // Prevent a user from ever having the ability to accept the base asset in case that becomes
+        // a contract - which may be used as a burned address similar to ETH 0x0000...
+        require(asset.value != BASE_ASSET_ID, CreationError::CannotUseBaseAsset);
```

Comment:
I've previously added `BASE_ASSET_ID` checks to my contracts in the case of something being burned, but have gone the route of instead using an `Option` and setting to `None`. For example in my NFT contract, the owner is now an `Option` and can be set to `None` so that any `Address` can be used including the `BASE_ASSET_ID` address. Will this be a reserved address?

Reply:
It was mentioned, in some conversation under a context I do not recall, that the 0th address may eventually be a contract (emphasise on the "may"). If that occurs and people treat it as a burner nonetheless then it's a good idea to block anything coming from there.

Reply:
Once the base becomes a flash loan contract then I don't know what to expect exactly. I don't want any nonsense happening so it's blocked

Reply:
I've already removed similar checks in other branches. Will likely do that here.